File: app/src/main/java/zoom/mcgars/com/zoom/MainActivity.java
package zoom.mcgars.com.zoom;

import android.content.Context;
import android.graphics.Bitmap;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;

import com.nostra13.universalimageloader.cache.memory.impl.UsingFreqLimitedMemoryCache;
import com.nostra13.universalimageloader.core.DisplayImageOptions;
import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.ImageLoaderConfiguration;
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener;

import java.util.ArrayList;
import java.util.List;

import mcgars.com.zoomimage.ZoomImageController;
import mcgars.com.zoomimage.adapter.ZoomPhotoPagerAdapter;
import mcgars.com.zoomimage.ui.Displayer;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private ZoomImageController zoomImageController;
    private ImageView ivTest;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initImageLoader(MainActivity.this);

        ivTest = (ImageView) findViewById(R.id.ivTest);

        zoomImageController = new ZoomImageController(this, new Displayer() {
            // load fullscreen image
            @Override
            public void displayImage(ZoomPhotoPagerAdapter.IPhoto photo, ZoomPhotoPagerAdapter.ViewHolder v) {
                displayImg(photo, v.image, v);
            }

            // load preview image
            @Override
            public void displayImage(ZoomPhotoPagerAdapter.IPhoto photo, ImageView v) {
                displayImg(photo, v, null);
            }

            @Override
            public void cancel(ImageView v) {
                ImageLoader.getInstance().cancelDisplayTask(v);
            }
        });


        ImageLoader.getInstance().displayImage("https://graceologydotcom.files.wordpress.com/2014/04/perma-glory.jpg", ivTest);
        ivTest.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        List<ZoomPhotoPagerAdapter.IPhoto> list = new ArrayList<>();
        list.add(new ZoomPhotoPagerAdapter.Photo(
                "https://graceologydotcom.files.wordpress.com/2014/04/perma-glory.jpg",
                "https://graceologydotcom.files.wordpress.com/2014/04/perma-glory.jpg"
        ));
        zoomImageController.setPhotos(ivTest, list);
        zoomImageController.enter(0);
    }

    /**
     * Display
     * @param photo
     * @param imageView
     * @param v
     */
    private void displayImg(ZoomPhotoPagerAdapter.IPhoto photo, ImageView imageView, final ZoomPhotoPagerAdapter.ViewHolder v) {
        ImageLoader.getInstance().displayImage(photo.getOriginal(), imageView, new SimpleImageLoadingListener() {
            @Override
            public void onLoadingComplete(String imageUri, View view, Bitmap loadedImage) {
                if(v!=null)
                    v.loadComplete();
            }
        });
    }

    /**
     * Init loader images
     * @param c
     * @return
     */
    public static ImageLoader initImageLoader(Context c) {
        if (ImageLoader.getInstance().isInited())
            return ImageLoader.getInstance();

        ImageLoaderConfiguration config = new ImageLoaderConfiguration.Builder(c)
                .defaultDisplayImageOptions(getImageLoaderOptions())
                .diskCacheSize(50 * 1024 * 1024)
                .diskCacheFileCount(100)
                .memoryCache(new UsingFreqLimitedMemoryCache(2 * 1024 * 1024)) // 2 Mb
                .build();

        ImageLoader.getInstance().init(config);
        return ImageLoader.getInstance();
    }

    public static DisplayImageOptions getImageLoaderOptions() {
        return getImageLoaderOptionsBuilder().build();
    }

    public static DisplayImageOptions.Builder getImageLoaderOptionsBuilder() {
        return new DisplayImageOptions.Builder()
                .cacheInMemory(true)
                .cacheOnDisk(true)
                .resetViewBeforeLoading(false);
    }

}

File: app/src/main/java/zoom/mcgars/com/zoom/MainActivity.kt
package zoom.mcgars.com.zoom

import android.content.Context
import android.graphics.Bitmap
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView

import com.nostra13.universalimageloader.cache.memory.impl.UsingFreqLimitedMemoryCache
import com.nostra13.universalimageloader.core.DisplayImageOptions
import com.nostra13.universalimageloader.core.ImageLoader
import com.nostra13.universalimageloader.core.ImageLoaderConfiguration
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener

import mcgars.com.zoomimage.ZoomImageController
import mcgars.com.zoomimage.adapter.ZoomViewHolder
import mcgars.com.zoomimage.fabric.AnimatorBuilder
import mcgars.com.zoomimage.model.IPhoto
import mcgars.com.zoomimage.model.Photo
import mcgars.com.zoomimage.ui.Displayer
import mcgars.com.zoomimage.ui.UiContainer

class MainActivity : AppCompatActivity(), View.OnClickListener {

    private lateinit var zoomImageController: ZoomImageController

    private val urlImages = listOf<IPhoto>(
            Photo(
                    "http://images4.fanpop.com/image/photos/23100000/-Nature-god-the-creator-23175770-670-446.jpg",
                    "http://fantasyartdesign.com/free-wallpapers/imgs/mid/decktop-Dino-game-m.jpg"
            ),
            Photo(
                    "http://fantasyartdesign.com/free-wallpapers/imgs/mid/decktop-Dino-game-m.jpg",
                    "http://images4.fanpop.com/image/photos/18200000/Lovely-nature-god-the-creator-18227423-500-333.jpg"
            ),
            Photo(
                    "http://images4.fanpop.com/image/photos/18200000/Lovely-nature-god-the-creator-18227423-500-333.jpg",
                    "http://images4.fanpop.com/image/photos/18200000/Lovely-nature-god-the-creator-18227423-500-333.jpg"
            )
    )

    private var views = mutableListOf<ImageView>()

    private val container: ViewGroup by lazy { findViewById<ViewGroup>(R.id.container) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        zoomImageController = ZoomImageController(UiContainer(findViewById(R.id.rootView)), object : Displayer {
            // load fullscreen image
            override fun displayImage(photo: IPhoto?, v: ZoomViewHolder) {
                displayImg(photo ?: return, v)
            }

            override fun cancel(imageView: ImageView) {
                ImageLoader.getInstance().cancelDisplayTask(imageView)
            }
        })

        initImageLoader(this)

        fitImages()

        //        ImageLoader.getInstance().displayImage("https://graceologydotcom.files.wordpress.com/2014/04/perma-glory.jpg", ivTest);
    }

    private fun fitImages() {
        views.add(findViewById<View>(R.id.ivTest) as ImageView)
        views.add(findViewById<View>(R.id.ivTest2) as ImageView)
        views.add(findViewById<View>(R.id.ivTest3) as ImageView)

        views.forEachIndexed { index, imageView ->
            imageView.setOnClickListener(this)
            ImageLoader.getInstance().displayImage(urlImages[index].preview, imageView)
        }
    }

    override fun onClick(v: View) {
        val position = container.indexOfChild(v)
        zoomImageController.setPhotos(AnimatorBuilder.from(views), urlImages)
        zoomImageController.show(position)
    }

    override fun onBackPressed() {
        if (zoomImageController.onBackPressed())
            return
        super.onBackPressed()
    }

    override fun onDestroy() {
        super.onDestroy()
        zoomImageController.onDestroy()
    }

    /**
     * Display
     * @param photo
     * @param v
     */
    private fun displayImg(photo: IPhoto, v: ZoomViewHolder) {
        ImageLoader.getInstance().displayImage(photo.original, v.image, object : SimpleImageLoadingListener() {

            override fun onLoadingStarted(imageUri: String?, view: View?) {

            }

            override fun onLoadingComplete(imageUri: String?, view: View?, loadedImage: Bitmap?) {
//                v.loadComplete()
            }
        })
    }

    companion object {

        /**
         * Init loader images
         * @param c
         * @return
         */
        fun initImageLoader(c: Context): ImageLoader {
            if (ImageLoader.getInstance().isInited)
                return ImageLoader.getInstance()

            val config = ImageLoaderConfiguration.Builder(c)
                    .defaultDisplayImageOptions(imageLoaderOptions)
                    .diskCacheSize(50 * 1024 * 1024)
                    .diskCacheFileCount(100)
                    .memoryCache(UsingFreqLimitedMemoryCache(2 * 1024 * 1024)) // 2 Mb
                    .build()

            ImageLoader.getInstance().init(config)
            return ImageLoader.getInstance()
        }

        val imageLoaderOptions: DisplayImageOptions
            get() = imageLoaderOptionsBuilder.build()

        val imageLoaderOptionsBuilder: DisplayImageOptions.Builder
            get() = DisplayImageOptions.Builder()
                    .cacheInMemory(true)
                    .cacheOnDisk(true)
                    .resetViewBeforeLoading(false)
    }

}
